http_server/handlers.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"html/template"
+	"crypto/rand"
+	"fmt"
+	"io"
+	"log"
+	"math/big"
+	"net/http"
+)
+
+// TODO(maramihali22): use template/html
+
+func (data *serviceData) welcomeHandler(w http.ResponseWriter, r *http.Request){
+	names, ok := r.URL.Query()["name"]
+	if !ok || len(names[0]) < 1 {
+		http.Error(w, "400 Bad Request", http.StatusBadRequest)
+		return
+	}
+	t, err := template.New("welcome").Parse(`{{define "T"}} Hello, {{.}}!{{end}}`)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+	// Assume the first query parameter correspond to the user's name
+	name := names[0]
+	err = t.ExecuteTemplate(w, "T", name)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+}
+
+func (data *serviceData) signUpHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		data.singUpGetHandlers(w, r)
+	case "POST":
+		data.signUpPostHandler(w, r)
+	default:
+		http.Error(w, "405 Method Not Allowed", http.StatusMethodNotAllowed)
+	}
+}
+
+func (data *serviceData) singUpGetHandlers(w http.ResponseWriter, r *http.Request) {
+	t := template.Must(template.ParseFiles("forms.html"))
+	err := t.Execute(w,struct{AccountExists bool}{false})
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)	
+	}
+}
+
+func (data *serviceData) signUpPostHandler(w http.ResponseWriter, r *http.Request) {	
+	data.mtx.Lock()
+	defer data.mtx.Unlock()
+
+	if err := r.ParseForm(); err != nil {
+		http.Error(w, "Failed to parse the form: %v. ", http.StatusInternalServerError)
+		return
+	}
+
+	u := userData{
+        username: r.FormValue("username"),
+        email: r.FormValue("email"),
+        password: r.FormValue("password"),
+	}
+
+	confirmPassword := r.FormValue("confirmPassword")
+
+	if u.password != confirmPassword {
+		http.Error(w, "Submitted passwords differ.", http.StatusBadRequest)
+		return
+	}
+
+	if _, taken := data.userMap[u.username]; taken {
+		http.Error(w, "Username already taken, please try again", http.StatusBadRequest)
+		return
+	}
+
+	data.userMap[u.username] = u
+
+	fmt.Fprintf(w, "Successfully signed up, %s! Please sign in", data.userMap[u.username].username)
+}
+
+func (data *serviceData) signInHandler(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "GET":
+		data.signInGetHandler(w, r)
+	case "POST":
+		data.signInPostHandler(w, r)
+	default:
+		http.Error(w, "405 Method Not Allowed", http.StatusMethodNotAllowed)
+	}
+
+}
+
+func (data *serviceData) signInGetHandler(w http.ResponseWriter, r *http.Request) {
+	
+	t := template.Must(template.ParseFiles("forms.html"))
```

Comment:
It is fine to have a template as a global, as it's read-only and you'll parse it just once.